spec/models/site_spec.rb

```diff
@@ -1,188 +1,348 @@
+        expect(site).to receive(:update_long_term_performance)
+
+        report = site.send(:generate_proxy_report, proxy)
+
+        expect(report.times_succeeded).to eq former_successes
+        expect(report.times_failed).to eq former_failures
+        expect(
+          @redis.zscore( site.proxy_successes.key, proxy.id )
+        ).to eq(0)
+        expect(
+          @redis.zscore( site.proxy_failures.key, proxy.id )
+        ).to eq(0)
+      end
+    end
+  end
+
+  describe 'PerformanceReport' do
+    it 'calculates the total number of successes + failures' do
+      report = Site::PerformanceReport.new(1, 6)
+      expect(report.total).to eq 7
+    end
+  end
+
+  describe 'performance analysis' do
+    before :each do
+      expect(site).to receive(:request_more_proxies)
+    end
+
+    it "should run a performance analysis on all proxies" do
+      2.times.each do |i|
+        proxy = create(:proxy, :port => i)
+        proxy.sites << site
+        proxy.save
+      end
+      expect(site).to receive(:disable_proxy_if_bad).twice
+
+      site.global_performance_analysis!
+    end
+
+    it "should run a performance analysis on a single proxy" do
+      expect(site).to receive(:disable_proxy_if_bad)
+      site.proxy_performance_analysis! proxy
+    end
+  end
+
+  context '#disable_proxy_if_bad' do
+
+    it "should not disable a successful proxy" do
+      report = Site::PerformanceReport.new(10, 1)
+      expect(site).to receive(:generate_proxy_report).and_return(report)
+      expect(site).to receive(:success_ratio_threshold).and_return(0.25)
+      expect(site).to receive(:disable_proxy).never
+
+      site.send(:disable_proxy_if_bad, proxy)
+    end
+
+    it "should disable an unsuccessful proxy" do
+      report = Site::PerformanceReport.new(10, 100)
+      expect(site).to receive(:generate_proxy_report).and_return(report)
+      expect(site).to receive(:success_ratio_threshold).and_return(0.25)
+      expect(site).to receive(:disable_proxy)
+
+      site.send(:disable_proxy_if_bad, proxy)
+    end
+  end
+
+  context '#request_more_proxies' do
+    it "does nothing if we don't need more proxies" do
+      expect(site).to receive(:num_proxies).and_return(5)
+      expect(site).to receive(:min_proxies).and_return(5)
+      expect(ProxyRequestor).to receive(:new).never
+
+      site.send(:request_more_proxies)
+    end
+
+    it "requests more proxies if we need more" do
+      expect(site).to receive(:num_proxies).and_return(4)
+      expect(site).to receive(:min_proxies).and_return(5)
+      requestor = double('requestor', :run => double)
+      expect(ProxyRequestor).to receive(:new).and_return(requestor)
+
+      site.send(:request_more_proxies)
+    end
+  end
+
+  context '#success_ratio_threshold' do
+    it 'retrieves the success ratio threshold config as a float' do
+      conf = double('conf', :[] => "0.25")
+      expect(Zartan::Config).to receive(:new).and_return(conf)
```

Comment:
The test environment has its own redis database separate from the other environments, so you can safely do this without doubles:
```ruby
Zartan::Config.new[:success_ratio_threshold] = 0.25
```